pkg/metrics/tls.go

```diff
@@ -81,12 +114,79 @@ func NewTLSConfigurerFromEnv() TLSConfigurer {
 
 // WatchForChanges watches for changes of the server TLS certificate files and the client CA config map.
 func (t *TLSConfigurerImpl) WatchForChanges() {
+	// Watch for changes of server TLS certificate.
+	certwatch.WatchCertDir(t.certDir, t.getCertificateFromDirectory, t.updateCertificate)
+
+	// Watch for changes of client CA.
+	go t.k8sWatcher.Watch(context.Background(), t.clientCANamespace, t.clientCAConfigMap)
 }
 
 // TLSConfig returns the current TLS config.
 func (t *TLSConfigurerImpl) TLSConfig() (*tls.Config, error) {
 	if t == nil {
 		return nil, nil
 	}
-	return nil, nil
+	return t.tlsConfigHolder.TLSConfig()
+}
+
+func (t *TLSConfigurerImpl) getCertificateFromDirectory(dir string) (*tls.Certificate, error) {
+	certFile := filepath.Join(dir, env.TLSCertFileName)
+	if exists, err := fileutils.Exists(certFile); err != nil || !exists {
+		if err != nil {
+			log.Warnw("Error checking if monitoring TLS certificate file exists", zap.Error(err))
+			return nil, err
+		}
+		log.Infof("Monitoring TLS certificate file %q does not exist. Skipping", certFile)
+		return nil, nil
+	}
+
+	keyFile := filepath.Join(dir, env.TLSKeyFileName)
+	if exists, err := fileutils.Exists(keyFile); err != nil || !exists {
+		if err != nil {
+			log.Warnw("Error checking if monitoring TLS key file exists", zap.Error(err))
+			return nil, err
+		}
+		log.Infof("Monitoring TLS key file %q does not exist. Skipping", keyFile)
+		return nil, nil
+	}
+
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return nil, errors.Wrap(err, "loading monitoring certificate failed")
+	}
+	cert.Leaf, err = x509.ParseCertificate(cert.Certificate[0])
+	if err != nil {
+		return nil, errors.Wrap(err, "parsing leaf certificate failed")
+	}
+	return &cert, nil
+}
+
+func (t *TLSConfigurerImpl) updateCertificate(cert *tls.Certificate) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
```

Comment:
super nit: theoretically only need RLock on `UpdateTLSConfig` but since this doesn't have a high load its not required at all.